model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,849 @@
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+        return origEntity -> new MapAuthenticatedClientSessionAdapter(session, realm, client, userSession, registerEntityForChanges(origEntity)) {
+            @Override
+            public void detachFromUserSession() {
+                //userSession.removeAuthenticatedClientSessions(Collections.singleton(client.getId()));
+
+                this.userSession = null;
+
+                clientSessionTx.delete(entity.getId());
+            }
+        };
+    }
+
+    private Function<MapUserLoginFailureEntity, UserLoginFailureModel> userLoginFailureEntityToAdapterFunc(RealmModel realm) {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+        return origEntity -> new MapUserLoginFailureAdapter(session, realm, registerEntityForChanges(origEntity));
+    }
+
+    private MapUserSessionEntity registerEntityForChanges(MapUserSessionEntity origEntity) {
+        MapUserSessionEntity res = userSessionTx.read(origEntity.getId(), id -> Serialization.from(origEntity));
+        userSessionTx.updateIfChanged(origEntity.getId(), res, MapUserSessionEntity::isUpdated);
+        return res;
+    }
+
+    private MapAuthenticatedClientSessionEntity registerEntityForChanges(MapAuthenticatedClientSessionEntity origEntity) {
+        MapAuthenticatedClientSessionEntity res = clientSessionTx.read(origEntity.getId(), id -> Serialization.from(origEntity));
+        clientSessionTx.updateIfChanged(origEntity.getId(), res, MapAuthenticatedClientSessionEntity::isUpdated);
+        return res;
+    }
+
+    private MapUserLoginFailureEntity registerEntityForChanges(MapUserLoginFailureEntity origEntity) {
+        MapUserLoginFailureEntity res = userLoginFailureTx.read(origEntity.getId(), id -> Serialization.from(origEntity));
+        userLoginFailureTx.updateIfChanged(origEntity.getId(), res, MapUserLoginFailureEntity::isUpdated);
+        return res;
+    }
+
+    @Override
+    public AuthenticatedClientSessionModel createClientSession(RealmModel realm, ClientModel client, UserSessionModel userSession) {
+        MapAuthenticatedClientSessionEntity entity =
+                new MapAuthenticatedClientSessionEntity(UUID.randomUUID(), realm.getId(), client.getId(), false);
+
+        if (clientSessionTx.read(entity.getId()) != null) {
+            throw new ModelDuplicateException("Client session exists: " + entity.getId());
+        }
+
+        LOG.tracef("createClientSession(%s, %s, %s)%s", realm, client, userSession, getShortStackTrace());
+
+        clientSessionTx.create(entity.getId(), entity);
+
+        MapUserSessionEntity userSessionEntity = getUserSessionById(UUID.fromString(userSession.getId()));
+        userSessionEntity.addAuthenticatedClientSession(client.getId(), entity.getId());
```

Comment:
Fixed.

Reply:
Actually I had to revert this change:
`MapUserSessionEntity userSessionEntity = registerEntityForChanges(getUserSessionById(UUID.fromString(userSession.getId())));`

back to:
`MapUserSessionEntity userSessionEntity = getUserSessionById(UUID.fromString(userSession.getId()));`

to avoid these test failures:
```
UserSessionProviderTest.testTransientUserSession:402->lambda$testTransientUserSession$8:410->assertSession:813 array lengths differed, expected.length=1 actual.length=0; arrays first differed at element [0]; expected:<test-app> but was:<end of array>
ServiceAccountTest.userInfoForServiceAccountWithoutRefreshTokenImpl:499 expected:<service-account-service-account-cl> but was:<null>
```
Or did you mean something else?

Reply:
I am afraid we need to fix the failed test in some other way than removing the `registerEntityForChanges` call. Currently, you are changing the MapUserSessionEntity directly when this code is executed and we need to make sure the change is reflected in storage on the transaction commit. Current implementation can result in some inconsistent state when some other transaction search for clientSessionId that is not yet created in the clientSessionStorage.

Reply:
I figured it out. I was trying to put a transient session into the map store which should be stored only directly in `MapUserSessionProvider` map.